node/src/isolate/external.cpp

```diff
@@ -0,0 +1,200 @@
+        on<protocol::choke>(std::bind(&spawn_dispatch_t::on_choke, this));
+        on<protocol::error>(std::bind(&spawn_dispatch_t::on_error, this, std::placeholders::_1, std::placeholders::_2));
+    }
+
+    void on_chunk(std::string chunk) {
+        if(chunk.empty()) {
+            spawn_handle->on_ready();
+        } else {
+            spawn_handle->on_data(std::move(chunk));
+        }
+
+    }
+
+    void on_choke() {
+        spawn_handle->on_terminate(std::error_code(), "");
+    }
+
+    void on_error(std::error_code ec, const std::string& msg) {
+        spawn_handle->on_terminate(ec, msg);
+    }
+
+    std::shared_ptr<api::spawn_handle_base_t> spawn_handle;
+};
+
+struct spawn_cancellation_t :
+public api::cancellation_t
+{
+    spawn_cancellation_t(io::upstream_ptr_t _stream) :
+        stream(std::move(_stream))
+    {}
+
+    ~spawn_cancellation_t() {
+        cancel();
+    }
+
+    virtual
+    void
+    cancel() {
+        if(cancelled.test_and_set()) {
+            stream->send<io::isolate_spawned::kill>();
+        }
+    }
+
+    io::upstream_ptr_t stream;
+    std::atomic_flag cancelled;
+};
+
+struct external_t::inner_t :
+    public std::enable_shared_from_this<inner_t>
+{
+    context_t& context;
+    asio::io_service& io_context;
+    std::unique_ptr<asio::ip::tcp::socket> socket;
+    std::shared_ptr<session_t> session;
+    const std::string name;
+    const dynamic_t args;
+    std::unique_ptr<logging::logger_t> log;
+
+    inner_t(context_t& _context, asio::io_service& _io_context, const std::string&, const dynamic_t& _args) :
+        context(_context),
+        io_context(_io_context),
+        socket(new asio::ip::tcp::socket(io_context)),
+        args(_args),
+        log(context.log("universal isolate"))
+    {
+
+    }
+
+    void connect() {
+        auto ep = args.as_object().at("endpoint", dynamic_t::empty_object).as_object();
+        tcp::endpoint endpoint(
+            asio::ip::address::from_string(ep.at("host", "127.0.0.1").as_string()),
+            static_cast<unsigned short>(ep.at("port", 29042u).as_uint())
+        );
+
+        auto self_shared = shared_from_this();
+        socket->async_connect(endpoint, [=](const std::error_code& ec) {
+            if (!ec) {
+                session = self_shared->context.engine().attach(std::move(socket), nullptr);
+            }
+        });
+    }
+
+};
+
+external_t::external_t(context_t& context, asio::io_service& io_context, const std::string& _name, const dynamic_t& _args) :
+    isolate_t(context, io_context, _name, _args),
+    inner(new inner_t(context, io_context, _name, _args))
+{
+    io_context.post(std::bind(&inner_t::connect, inner));
+}
+
+std::unique_ptr<api::cancellation_t>
+external_t::async_spool(std::shared_ptr<api::spool_handle_base_t> handler) {
+    if(!inner->session) {
+        throw cocaine::error_t(error::dispatch_errors::not_connected, "universal isolation daemon is not connected");
```

Comment:
How do you can guarantee not calling this method before the connection is established (becase it's in progress for example)? It will be more useful to delay erroring either until we sure that connection can't be established.

Using futures would help a lot, but we haven't them.